### DISCUSSION TITLE
Automatically iterating through each AOI for grid generation rather than having the users handle it manually

### DISCUSSION THREAD

**Discussion by jtmiclat** [thinkingmachines/geowrangler#63] [New Feature Ideas]
Based on the @joshuacortez  suggested https://github.com/thinkingmachines/geowrangler/pull/55#issuecomment-1170880281
Some questions:
- Should we iterate on each AOI and deduplicate by default? If yes how do we know when to do and when just do the naive version 
- Should this be a separate function or param or just not configurable?

**Comment by jtmiclat** [comment]
A possible option is to check the [unary union](https://geopandas.org/en/stable/docs/reference/api/geopandas.GeoSeries.unary_union.html) of the AOIs and if it is a `Polygon` do the native method and if `MultiPolygon` do the iterative method.

**Reply by joshuacortez** [reply]
Hmm could we just make it iterative by default?

**Reply by jtmiclat** [reply]
Iterative is an improvement for AOIs that are non-adjacent but i believe if the aoi are adjacent it would be more performance to not iterate through them.

**Reply by jtmiclat** [reply]
We might need benchmark and compare these scenarios.

**Comment by butchtm** [comment]
Would it be possible to parallelize this by running the grid computation for each aoi independently in parallel and deduplicating the resulting rows in the end? Maybe by using  something like [dask-pandas](https://docs.dask.org/en/stable/generated/dask.dataframe.from_pandas.html)? Of course this should be completely optional and useful only when doing fine-grained grids covering large areas.

**Comment by jtmiclat** [comment]
Just a heads up, we are using this logic to handle generation of grids for h3 and bing tiles
https://github.com/thinkingmachines/geowrangler/blob/4325da086c73718c0204b9875608e6035be4eb0a/geowrangler/grids.py#L161-L166

**Reply by joshuacortez** [reply]
Are we planning to use this for a list of custom AOIs too? I'm ok if the extra time to do a `unary_union` isn't unbearable

**Reply by jtmiclat** [reply]
Yup! Will measure the performance impact of this methodology and compare the other one.

**Reply by joshuacortez** [reply]
awesome thanks!

**Reply by jtmiclat** [reply]
I made an initial test. For unary_union to make 5km grids off of `data/region3.json` it took
```
CPU times: user 918 ms, sys: 12 ms, total: 930 ms
Wall time: 930 ms
```
while the current way took no iteration took 
```
CPU times: user 38.9 s, sys: 57.8 ms, total: 38.9 s
Wall time: 41.1 s
```
Will bench mark iterating each aoi vs iterating unary_union!

**Reply by jtmiclat** [reply]
Benchmark based on gpkg mentioned here   https://github.com/thinkingmachines/geowrangler/pull/55#issuecomment-1169536044
```
Iterating through each aoi (300 rows) 
CPU times: user 2min 7s, sys: 1.68 s, total: 2min 8s
Wall time: 2min 10s
Iterating through each unary union (300 rows) 
CPU times: user 42.2 s, sys: 1.15 s, total: 43.3 s
Wall time: 43.6 s

Iterating through each aoi (4k rows)
CPU times: user 52min 34s, sys: 23.3 s, total: 52min 58s
Wall time: 52min 38s

Iterating through each unary union(4k rows)
CPU times: user 41min 17s, sys: 36.5 s, total: 41min 54s
Wall time: 41min 56s
```

**Reply by jtmiclat** [reply]
Created #90 to fully implement this!

**Reply by joshuacortez** [reply]
Cool, so the `unary_union` shaves off time even if the polygons are mostly adjacent or not

**Comment by joshuacortez** [comment]
Closing since this is already addressed: using `unary_union` in Square Grid Generator and Bing Tile Grid Generator
